crypto/evp/p_lib.c

```diff
@@ -80,66 +81,141 @@ int EVP_PKEY_save_parameters(EVP_PKEY *pkey, int mode)
 
 int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from)
 {
-    if (to->type == EVP_PKEY_NONE) {
-        if (EVP_PKEY_set_type(to, from->type) == 0)
+    int is_provided = 0;
+
+    if (from->ameth == NULL
+        && evp_pkey_make_provided((EVP_PKEY *)from, NULL, NULL, NULL, -1)) {
+        /*
+         * It's fine if the destination hasn't cached anything yet, that
+         * simply makes it a blank page to be filled in with the
+         * evp_keymgmt_copy() call below.
+         * That call will also detect |from->pkeys[0].keymgmt == NULL|,
+         * so we don't bother making an error here.
+         */
+        if (to->pkeys[0].keymgmt != NULL
+            && from->pkeys[0].keymgmt != NULL
+            && (EVP_KEYMGMT_number(to->pkeys[0].keymgmt)
+                != EVP_KEYMGMT_number(from->pkeys[0].keymgmt))) {
```

Comment:
So if the keymgmt instances are from different libctxs won't this comparison cause a problem (because the numbers might be different for the same keymgmt in different libctxs)?

Reply:
Ouch...  Okie, will give this another thought